File: Sources/Server/Sources/Game/Genres/Wave/WaveRoomService.ts
import type { Biome } from "../../../../../Shared/Biome";
import { WaveRoomVisibleState } from "../../../../../Shared/WaveRoom";
import type { WaveRoomCode} from "../../../../../Shared/WaveRoomCode";
import { generateRandomWaveRoomCode } from "../../../../../Shared/WaveRoomCode";
import { logger } from "../../../../Main";
import type { UserData } from "./WavePool";
import type { WaveRoomPlayerId } from "./WaveRoom";
import WaveRoom from "./WaveRoom";

export default class WaveRoomService {
    private waveRooms: WaveRoom[] = [];

    /**
     * Adds a player to an existing public wave room or creates a new one if none exists.
     */
    public joinPublicWaveRoom(userData: UserData, biome: Biome): WaveRoomPlayerId | false {
        if (!this.canUserDataAdded(userData)) {
            return false;
        }

        this.leaveCurrentWaveRoom(userData);

        let waveRoom = this.findPublicRoom(biome);

        if (!waveRoom) {
            return this.createWaveRoom(userData, biome);
        } else {
            return waveRoom.registerPlayer(userData.staticPlayerData);
        }
    }

    /**
     * Adds a player to a private wave room using a room code.
     */
    public joinWaveRoom(userData: UserData, code: WaveRoomCode): WaveRoomPlayerId | false {
        if (!this.canUserDataAdded(userData)) {
            return false;
        }

        this.leaveCurrentWaveRoom(userData);

        const room = this.findPrivateRoom(code);
        if (!room) {
            return false;
        }

        return room.registerPlayer(userData.staticPlayerData);
    }

    /**
     * Removes a player from their wave room, delete empty rooms if wave room is empty.
     */
    public leaveWaveRoom(id: WaveRoomPlayerId): boolean {
        for (const waveRoom of this.waveRooms) {
            if (waveRoom.unregisterPlayer(id)) {
                if (waveRoom.roomCandidates.length === 0) {
                    this.removeWaveRoom(waveRoom);
                }

                return true;
            }
        }

        return false;
    }

    /**
     * Removes a specific wave room from the manager.
     * 
     * @remarks
     * 
     * This release wavePool memory too.
     */
    public removeWaveRoom(waveRoom: WaveRoom) {
        waveRoom.releaseAllMemory();

        this.waveRooms.splice(this.waveRooms.indexOf(waveRoom), 1);

        logger.region(() => {
            using _guard = logger.metadata({ code: waveRoom.code });
            logger.info("Removed wave (room)");
        });
    }

    /**
     * Creates a new wave room with initial player.
     */
    public createWaveRoom(userData: UserData, biome: Biome): WaveRoomPlayerId | false {
        if (!this.canUserDataAdded(userData)) {
            return false;
        }

        this.leaveCurrentWaveRoom(userData);

        const waveRoom = new WaveRoom(biome, this.generateCode());
        this.waveRooms.push(waveRoom);

        logger.region(() => {
            using _guard = logger.metadata({ code: waveRoom.code });
            logger.info("Created wave room");
        });

        return waveRoom.registerPlayer(userData.staticPlayerData);
    }

    /**
     * Finds a public room with available slots for the specified biome.
     */
    private findPublicRoom(biome: Biome): WaveRoom | undefined {
        // Trying to find biome-specific room, if not found, try other biome
        const isPublic = (room: WaveRoom) => room.visible === WaveRoomVisibleState.Public && room.newPlayerAcceptable;

        return this.waveRooms.find(room => room.biome === biome && isPublic(room)) || this.waveRooms.find(room => isPublic(room));
    }

    /**
     * Finds a private room with available slots by room code.
     * 
     * @remarks
     * This don't actually find a private room, just find a room with the same code.
     */
    private findPrivateRoom(code: WaveRoomCode): WaveRoom | undefined {
        return this.waveRooms.find(room => room.code === code && room.newPlayerAcceptable);
    }

    /**
     * Finds the wave room that contains a specific player.
     */
    public findPlayerRoom(id: WaveRoomPlayerId): WaveRoom | null {
        for (const waveRoom of this.waveRooms) {
            const findedPlayer = waveRoom.roomCandidates.find(p => p.id === id);
            if (findedPlayer) {
                return waveRoom;
            }
        }

        return null;
    }

    /**
     * Generates a unique room code.
     */
    private generateCode(): WaveRoomCode {
        const randomId = generateRandomWaveRoomCode();
        if (!this.waveRooms.every((room) => room.code !== randomId)) {
            return this.generateCode();
        }
        
        return randomId;
    }

    /**
     * Remove player from wave room if player were in other wave room.
     * 
     * @remarks
     * The first time this code is executed, the room which player was in will no finded by find○○Room.
     * However, placing this at the beginning of the code means that the player will exit the player's current room even if they are not joining a room.
     * I don't know how this works in the original, but this maybe not correct.
     */
    private leaveCurrentWaveRoom(userData: UserData) {
        if (userData?.waveRoomClientId) this.leaveWaveRoom(userData.waveRoomClientId);
    }

    /**
     * Determine if user data is allowed to join/add.
     */
    private canUserDataAdded(userData: UserData): boolean {
        return !userData || !userData.staticPlayerData ||
            // Anti ws duplicate
            !new Set(this.waveRooms.map(waveRoom => waveRoom.wavePool.getAllClients().map(c => c.ws)).flat()).has(userData.staticPlayerData.ws);
    }

    /**
     * Get all wave room.
     */
    public get allWaveRoom(): WaveRoom[] {
        return this.waveRooms;
    }
}

File: Sources/goserver/internal/game/wave_room_service.go
package game

import (
	"slices"
	"sync"

	"flooooio/internal/native"
)

type WaveRoomService struct {
	waveRooms []*WaveRoom

	mu sync.RWMutex
}

var WrService = NewWaveRoomService()

func NewWaveRoomService() *WaveRoomService {
	return &WaveRoomService{
		waveRooms: make([]*WaveRoom, 0),
	}
}

// JoinPublicWaveRoom adds a player to an existing public wave room or creates a new one if none exists.
func (s *WaveRoomService) JoinPublicWaveRoom(pd *PlayerData, biome native.Biome) *WaveRoomPlayerId {
	// Place this before lock so doesnt double call lock
	s.LeaveCurrentWaveRoom(pd)

	s.mu.Lock()
	defer s.mu.Unlock()

	if !s.canPlayerJoin(pd) {
		return nil
	}

	// Try to find existing public room
	if room := s.findPublicRoom(biome); room != nil {
		return room.RegisterPlayer(pd.Sp)
	}

	// Return nil if not found
	return nil
}

// JoinWaveRoom adds a player to a private wave room using a room code.
func (s *WaveRoomService) JoinWaveRoom(pd *PlayerData, code WaveRoomCode) *WaveRoomPlayerId {
	// Place this before lock so doesnt double call lock
	s.LeaveCurrentWaveRoom(pd)

	s.mu.Lock()
	defer s.mu.Unlock()

	if !s.canPlayerJoin(pd) {
		return nil
	}

	room := s.findPrivateRoom(code)
	if room == nil {
		return nil
	}

	return room.RegisterPlayer(pd.Sp)
}

// removeWaveRoom removes a specific wave room from the manager.
func (s *WaveRoomService) removeWaveRoom(room *WaveRoom) {
	for i, r := range s.waveRooms {
		if r == room {
			room.Dispose()

			s.waveRooms = slices.Delete(s.waveRooms, i, i+1)

			return
		}
	}
}

// leaveWaveRoom removes a player from their wave room, deletes empty rooms.
func (s *WaveRoomService) leaveWaveRoom(id WaveRoomPlayerId) (ok bool) {
	for _, room := range s.waveRooms {
		if ok := room.DeregisterPlayer(id); ok {
			if len(room.candidates) == 0 {
				// Remove empty room
				s.removeWaveRoom(room)
			}

			return true
		}
	}

	return false
}

// LeaveCurrentWaveRoom removes player from wave room if player was in other wave room.
func (s *WaveRoomService) LeaveCurrentWaveRoom(pd *PlayerData) (ok bool) {
	s.mu.Lock()
	defer s.mu.Unlock()

	if pd.WrPId != nil {
		return s.leaveWaveRoom(*pd.WrPId)
	}

	return false
}

// NewPublicWaveRoom creates a new public wave room with initial player.
func (s *WaveRoomService) NewPublicWaveRoom(pd *PlayerData, biome native.Biome) *WaveRoomPlayerId {
	// Place this before lock so doesnt double call lock
	s.LeaveCurrentWaveRoom(pd)

	s.mu.Lock()
	defer s.mu.Unlock()

	if !s.canPlayerJoin(pd) {
		return nil
	}

	room := NewWaveRoom(biome, RoomVisibilityPublic)

	s.waveRooms = append(s.waveRooms, room)

	return room.RegisterPlayer(pd.Sp)
}

// FindPublicRoom finds a public room with available slots for the specified biome.
func (s *WaveRoomService) findPublicRoom(biome native.Biome) *WaveRoom {
	// First try to find room with matching biome
	for _, room := range s.waveRooms {
		if room.visibility == RoomVisibilityPublic &&
			room.biome == biome &&
			room.isNewPlayerRegisterable() {
			return room
		}
	}

	// If no matching biome found, return any public room
	for _, room := range s.waveRooms {
		if room.visibility == RoomVisibilityPublic &&
			room.isNewPlayerRegisterable() {
			return room
		}
	}

	return nil
}

// FindPrivateRoom finds a private room with available slots by room code.
func (s *WaveRoomService) findPrivateRoom(code WaveRoomCode) *WaveRoom {
	for _, room := range s.waveRooms {
		if room.code == code && room.isNewPlayerRegisterable() {
			return room
		}
	}

	return nil
}

// FindPlayerRoom finds the wave room that contains a specific player.
func (s *WaveRoomService) FindPlayerRoom(id WaveRoomPlayerId) *WaveRoom {
	s.mu.RLock()
	defer s.mu.RUnlock()

	for _, room := range s.waveRooms {
		for _, candidate := range room.candidates {
			if candidate != nil && candidate.Id == id {
				return room
			}
		}
	}

	return nil
}

// canPlayerJoin determines if a player is allowed to join/add.
func (s *WaveRoomService) canPlayerJoin(pd *PlayerData) bool {
	// Check if player is already in a room
	for _, room := range s.waveRooms {
		for _, candidate := range room.candidates {
			if candidate != nil && candidate.Conn == pd.Sp.Conn {
				return false
			}
		}
	}

	return true
}

func RemovePlayerFromService(pd *PlayerData) {
	if pd.WrPId != nil && pd.WPId != nil {
		wr := WrService.FindPlayerRoom(*pd.WrPId)

		if wr != nil {
			wp := wr.WavePool

			if wp != nil {
				player := wp.FindPlayer(*pd.WPId)

				if player != nil {
					DisposeBindings(wp, player)

					wp.RemovePlayer(*pd.WPId)

					pd.WPId = nil
				}
			}
		}
	}

	{ // This block should executed after because needs to use FindPlayerRoom
		// Dont care about result
		_ = WrService.LeaveCurrentWaveRoom(pd)

		pd.WrPId = nil
	}
}
